### PR TITLE
HV-1527 Separate Group and Path processed units in ValidationContext

### PR DESCRIPTION
**Pull request by marko-bekhta**
- https://hibernate.atlassian.net/browse/HV-1527

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 1]
```diff
@@ -91,9 +91,15 @@
 	private final Object executableReturnValue;
 
 	/**
-	 * The set of already processed unit of works. See {@link ProcessedUnit}.
+	 * The sets of already processed path meta constraint unit of works({@link BeanPathMetaConstraintProcessedUnit})
+	 * {@link }.
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: There's something wrong with the javadoc here.

+ I think you can remove "of works" and make unit a plural.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 2]
```diff
@@ -91,9 +91,15 @@
...
 	 */
-	private final Set<Object> processedUnits;
+	private final Set<BeanPathMetaConstraintProcessedUnit> processedPathUnits;
+
+	/**
+	 * The sets of already processed group unit of works({@link BeanGroupProcessedUnit})
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: Same here about the second remark.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 3]
```diff
@@ -611,13 +618,8 @@ private BeanGroupProcessedUnit(Object bean, Class<?> group) {
 
 		@Override
 		public boolean equals(Object o) {
-			if ( this == o ) {
-				return true;
-			}
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: Do we really gain something by removing this? If not I would vote for keeping it (and update the comment below).

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 4]
```diff
@@ -659,22 +661,17 @@ private BeanPathMetaConstraintProcessedUnit(Object bean, Path path, MetaConstrai
...
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: Same here.

Comment (marko-bekhta) [reply engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: I've checked. It don't really make a difference, results are almost identical. So I've added these back.

**Comment by gsmet** [comment]
Jenkins, retest this please

**Comment by gsmet** [comment]
OK, looks good now. I'll rework some of the comments and merge tomorrow morning.

Thanks!

**Comment by gsmet** [comment]
Merged, thanks!

**Pull request merged**